compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotter.kt

```diff
@@ -57,26 +63,66 @@ object ClassSnapshotter {
+        classes: List<ClassFileWithContentsProvider>,
+        granularity: ClassSnapshotGranularity,
         metrics: BuildMetricsReporter = DoNothingBuildMetricsReporter
     ): List<ClassSnapshot> {
-        val classesInfo: List<BasicClassInfo> = metrics.measure(BuildTime.READ_CLASSES_BASIC_INFO) {
-            classes.map { it.classInfo }
-        }
-        val inaccessibleClassesInfo: Set<BasicClassInfo> = metrics.measure(BuildTime.FIND_INACCESSIBLE_CLASSES) {
-            findInaccessibleClasses(classesInfo)
-        }
-        return classes.map {
-            when {
-                it.classInfo in inaccessibleClassesInfo -> InaccessibleClassSnapshot
-                it.classInfo.isKotlinClass -> metrics.measure(BuildTime.SNAPSHOT_KOTLIN_CLASSES) {
-                    snapshotKotlinClass(it, granularity)
+        val classSnapshots = mutableMapOf<ClassFileWithContentsProvider, ClassSnapshot>()
+        val inaccessibleClasses = mutableSetOf<ClassId>()
+
+        // Process the classes in descending order of their relative paths so that outer classes are processed first, which is needed for
+        // detecting transitively inaccessible classes -- see below.
+        // For example, if `com/example/A$B.class` has outer class `com/example/A.class`, we will process `com/example/A.class` first (the
+        // string `com/example/A.class` is higher than `com/example/A$B.class` because `.` is higher than `$`).
+        classes.sortedByDescending { it.classFile.unixStyleRelativePath }.forEach {
+            val clazz = metrics.measure(BuildTime.LOAD_CONTENTS_OF_CLASSES) {
+                it.loadContents()
+            }
+            val classSnapshot = when {
+                // We don't need to snapshot classes that are (directly or transitively) inaccessible.
+                // A class is transitively inaccessible if its outer class is (directly or transitively) inaccessible.
+                // The outer class (if present) should have been processed due to the way we sorted the given classes (see above).
+                //
+                // In the uncommon case that the outer class is not yet processed (i.e., when the outer class's path is not a prefix of this
+                // class's path) or if it can't be found in the given classes (this could happen with faulty jars), then we will not be able
+                // to confirm whether this class is transitively inaccessible. If we try to find and process the outer class on the fly, we
+                // will potentially need to load many more classes, and we will also have to cache them in memory to avoid loading them
+                // twice later, therefore consuming more memory (KT-57757). To keep things simple and avoid KT-57757, we'll just consider
+                // that the class is accessible and snapshot it (it's okay to snapshot more than required, but incorrect to snapshot less).
+                // Note that this case is unlikely to happen anyway.
```

Comment:
While this seems to address this use-case, I think we should still file a YT issue to switch to on-demand loading of classes. If we need B in order to process A, the class loading mechanism can provide this information. Currently, sorting of entries addresses the most common scenario, but if there are new cross-class requirements, this becomes difficult to maintain.

Also, the comment about the memory consumption does not really hold, in order to load A, you'll need to load the outer class parent chain. which should not be more than ~10 classes.

Reply:
> Also, the comment about the memory consumption does not really hold, in order to load A, you'll need to load the outer class parent chain. which should not be more than ~10 classes.

The key challenge is that we can only get a `ClassID` by reading the `.class` file's contents, not by its path, because the `$` character in the file path is ambiguous (`A$B.class` can either mean inner class `B` of outer class `A`, or it can mean top-level class `A$B` where `$` is part of its name).

So to find the outer class we will potentially need to look for + load all entries in the jar.

I've updated the algorithm a bit to look for outer classes, could you take another look to see if it's better?

-----------------------------------

(Btw, I tested on the 400 MB `ideaIC-2022.1.4/lib/app.jar` and got some numbers:

-  Directly inaccessible classes: 20241
- Transitively inaccessible classes: 0 (we never found a case where a class says it's accessible but its outer class is inaccessible)
- Kotlin classes: 9377
- Java classes: 47854

So I'll need to be careful not to increase code complexity too much for detecting "transitively inaccessible" classes.)